src/EditorFeatures/Core/Implementation/IntelliSense/Completion2/RoslynCompletionItemSource.cs

```diff
@@ -0,0 +1,201 @@
+                    }
+                    else
+                    {
+                        var itemFilter = new CompletionFilter(filter.DisplayText, filter.AccessKey.ToString(), imageService.GetImageId(filter.Tags.GetGlyph()));
+                        filterCache[filter.DisplayText] = itemFilter;
+                        result.Add(itemFilter);
+                    }
+                }
+            }
+
+            return result.ToImmutableArray();
+        }
+
+        public Task<object> GetDescriptionAsync(EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            item.Properties.TryGetProperty<object>("Description", out var description);
+            return Task.FromResult(description);
+        }
+
+        public void CustomCommit(ITextView view, ITextBuffer buffer, EditorCompletion.CompletionItem item, ITrackingSpan applicableSpan, char commitCharacter)
+        {
+            var service = GetCompletionService(buffer.CurrentSnapshot) as CompletionServiceWithProviders;
+
+            var roslynItem = item.Properties.GetProperty<RoslynCompletionItem>(RoslynItem); // We're using custom data we deposited in GetCompletionContextAsync
+            var triggerSnapshot = item.Properties.GetProperty<ITextSnapshot>(TriggerSnapshot);
+
+            var edit = buffer.CreateEdit();
+            var provider = service.GetProvider(roslynItem);
+            if (provider is ICustomCommitCompletionProvider c)
+            {
+                c.Commit(roslynItem, view, buffer, triggerSnapshot, null);
+            }
+            else
+            {
+                var document = buffer.GetRelatedDocuments().First();
+                var roslynChange = service.GetChangeAsync(document, roslynItem, commitCharacter, CancellationToken.None).Result;
+
+                // TODO: Editor to reapply inserted trigger after we commit
+                var ts = new SnapshotSpan(triggerSnapshot, roslynChange.TextChange.Span.ToSpan());
+                var mapped = ts.TranslateTo(buffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                edit.Replace(mapped.Span, roslynChange.TextChange.NewText);
+            }
+
+            edit.Apply();
+        }
+
+        public ImmutableArray<char> GetPotentialCommitCharacters()
+        {
+            return CommitChars;
+        }
+
+        public bool ShouldCommitCompletion(char typedChar, SnapshotPoint location)
+        {
+            return CommitChars.Contains(typedChar);
+        }
+
+        public bool ShouldTriggerCompletion(char edit, SnapshotPoint location)
+        {
+            var text = SourceText.From(location.Snapshot.GetText());
+            var service = GetCompletionService(location.Snapshot);
+            return service?.ShouldTriggerCompletion(text, location.Position, RoslynTrigger.CreateInsertionTrigger(edit)) ?? false;
+        }
+
+        private CompletionService GetCompletionService(ITextSnapshot snapshot)
```

Comment:
We make some assumptions earlier in the file that this is going to be a ServiceWithProviders--should we be more explicit about that here?